### PR TITLE
107 add system type

### LINKED ISSUE #105: Add a system type
As a user, I want to be able to include different system types by adding a system type.

Acceptance Criteria:
- [x] A system type can be added through the UI
- [x] Threshold values for the different recommendation boundaries can be chosen
- [x] Assets can be labeled of that new entered type when they are added to the application

*Issue #48 was divided in 2 (#48 is for remove the system type and this one is for adding a system type)

### LINKED ISSUE #107: UI - Create a "Add System Type" screen
Since we need our users to add system types in the application, we need to have the UI to allow such action.

Acceptance Criteria:
- [x] Create a mock-up for the 'Add System Type' Screen and update the documentation on Confluence
- [x] Create the Screen based on the mock-up. It should contain the following:
- Add a name - field
- Choose the threshold values for the different recommendation boundaries - field

### LINKED ISSUE #122: UI/Mock-up - Create a "Add System Type" screen Mock-up
Since we need our users to add system types in the application, we need to have the UI to allow such action.

Acceptance Criteria:
- [x]  Create a mock-up for the 'Add System Type' Screen and update the documentation on Confluence
- [x] Add a name - field
- [x] Choose the threshold values for the different recommendation boundaries - field

### PR DESCRIPTION
**Pull request by jeffbuonamici**
User now has the ability to add different system types. System types have a name, description, and parameters(threshold values). The user can add as many threshold values as they need by clicking the "+" button. Once the user clicks save, it gets added into the database.

This pull request also include issue #105. I thought #107 was a completing the whole functionality, however it was just for the UI.

### CODE REVIEW

**Comment by PaulMicu** [comment]
i dont see where in the UI you can add the system type from the main page, can you clarify, thanks

**Comment by jeffbuonamici** [comment]
@PaulMicu You can't right now because the page where you would access it is currently being implemented by @najimgh. Simply change "/Systems.fxml" in Main.java to "/AddSystemType.fxml" just to test it.

**Review by Shirwarsame — COMMENTED** [review]
Code works and is put into the database
System Type Name  and Threshold Name allows just numbers to be accepted
System information needs to be stored somewhere in the database

File: src/main/java/com/cbms/ui/controller/AddSystemTypeController.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import javafx.scene.text.Font;
+import javafx.scene.text.Text;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.ResourceBundle;
+
+public class AddSystemTypeController implements Initializable {
+
+    @FXML
+    private Button systemMenuButton;
```
Comment (JayChouteau) [inline src/main/java/com/cbms/ui/controller/AddSystemTypeController.java]: Change the naming to systemMenuBtn (to be consistent with the UI refactoring)

File: src/main/java/com/cbms/ui/controller/AddSystemTypeController.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    @FXML
+    private Button cancelButton;
```
Comment (JayChouteau) [inline src/main/java/com/cbms/ui/controller/AddSystemTypeController.java]: Change the naming to cancelBtn (to be consistent with the UI refactoring)

File: src/main/java/com/cbms/ui/controller/AddSystemTypeController.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    @FXML
+    private Button addThresholdBtn;
+    @FXML
+    private AnchorPane systemTypeInformation;
+    @FXML
+    private Button savebtn;
```
Comment (JayChouteau) [inline src/main/java/com/cbms/ui/controller/AddSystemTypeController.java]: -> saveBtn

File: src/main/resources/AddSystemType.fxml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                <font>
+                    <Font name="Segoe UI Bold" size="14.0" />
+                </font>
+                <graphic>
+                    <ImageView fitHeight="24.0" fitWidth="24.0" pickOnBounds="true" preserveRatio="true">
+                        <Image url="@imgs/outline_dvr_white_18dp.png" />
+                    </ImageView>
+                </graphic>
+            </Button>
+            <Button fx:id="exitMenuButton" contentDisplay="TOP" layoutX="10.0" layoutY="10.0" mnemonicParsing="false" prefHeight="74.0" prefWidth="74.0" style="-fx-background-radius: 0;" stylesheets="@css/style.css" text="Exit" textAlignment="CENTER" textFill="#f0f0f0" translateY="375.0">
```
Comment (JayChouteau) [inline src/main/resources/AddSystemType.fxml]: ->exitMenuBtn

File: src/main/resources/css/style.css [hunk 5]
```diff
@@ -25,6 +25,11 @@
 }
 
 #addSystemButton
```
Comment (JayChouteau) [inline src/main/resources/css/style.css:27]: ->addSystemButton

File: src/main/resources/AddSystemType.fxml [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+                <font>
+                    <Font name="Segoe UI Bold" size="14.0" />
+                </font>
+                <graphic>
+                    <ImageView fitHeight="24.0" fitWidth="24.0" pickOnBounds="true" preserveRatio="true">
+                        <image>
+                            <Image url="@imgs/outline_exit_to_app_white_18dp.png" />
```
Comment (JayChouteau) [inline src/main/resources/AddSystemType.fxml:34]: My intellij tells me that I cannot resolve this file. Was this added to our project? I don't see it in the imgs folder.

File: src/main/java/com/cbms/ui/controller/AddSystemTypeController.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                uiUtilities.changeScene(mouseEvent, "/Systems");
+            }
+        });
+        addThresholdBtn.setOnMouseClicked(new EventHandler<MouseEvent>() {
+            @Override
+            public void handle(MouseEvent mouseEvent) {
+                addThresholdForm();
+            }
+        });
+        savebtn.setOnMouseClicked(new EventHandler<MouseEvent>() {
```
Comment (JayChouteau) [inline src/main/java/com/cbms/ui/controller/AddSystemTypeController.java]: Cap on value
My understanding is that the threshold value is a % of the RUL, meaning that there should be a min value (0) and a max value (100). This should be enforced in the controller.

File: src/main/resources/AddSystemType.fxml [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                    <Font name="Segoe UI Bold" size="14.0" />
+                </font>
+            </Text>
+            <Text fill="#0c072e" layoutX="26.0" layoutY="118.0" strokeType="OUTSIDE" strokeWidth="0.0" text="System Type Description">
+                <font>
+                    <Font name="Segoe UI Bold" size="14.0" />
+                </font>
+            </Text>
+            <TextArea layoutX="200.0" layoutY="100.0" prefHeight="63.0" prefWidth="350.0" />
+            <TextField fx:id="thresholdName1" layoutX="200.0" layoutY="175.0" prefHeight="25.0" prefWidth="190.0" />
```
Comment (JayChouteau) [inline src/main/resources/AddSystemType.fxml:68]: 1) Threshold names and values
I feel like UI wise, the labels we put for the threshold name and value are maybe too abstract for users. I feel like changing "Threshold name" and "value" to "Recommendation label" and "Threshold value" might be more explicit for a user.
2) Threshold description
I feel like we should have a description (text) on the UI that explains what the threshold name and values represent for the user to enter information correctly. Either a plain description or a small help button.

**Review by JayChouteau — CHANGES_REQUESTED** [review]
Overall, very good work on the issue. The UI is clean and the code structure follows what was built. I left a few small comments directly on the code. Some of the other feedback:
1) Number of thresholds
This is more about how we want to deal with recommendation labels, but do we want to limit the number of threshold labels that can be assigned to an asset type? Do we want to offer a selection instead of having the user enter his own? This doesn't block the pr or issue but is more food for thought for the team.
2) Mimosa architecture
As we move on and add to our item package, i am wondering if this follows the mimosa architecture. We don't really need to stick to it, but I guess it was our model. Roy might be better to help with this. Plus, this comment doesn't stop the merging of this pull request.

**Comment by jeffbuonamici** [comment]
@JayChouteau I made the variable name changes. The rest of your concerns can be changed once we have more clarification on these issues.

**Review by JayChouteau — APPROVED** [review]

**Pull request merged**